doc/RawSyscallsDesign.md

```diff
@@ -0,0 +1,196 @@
+   clobber list. It would be inefficient to share an `asm!` call between Yield
+   and another system call.
+1. Exit does not return. This initially seems to be an important distinction,
+   but it is not. `Syscalls` can use `unreachable_unchecked!` to tell the
+   compiler it doesn't exit, and it should be obvious the return values are all
+   unused.
+1. The compiler is unable to optimize away unused arguments. For example,
+   Memop's "get process RAM start address" operation only needs `r0` set, while
+   Memop's "break" operation needs both `r0` and `r1` set. If our inline
+   assembly calls "get process RAM start address" but sets both `r0` and `r1`,
+   the compiler doesn't know that `r1` will be ignored so setting that register
+   will not be optimized away. Therefore we want to set the minimum number of
+   argument registers possible.
+1. The cost of specifying unused return registers is only that of unnecessarily
+   marking a register as clobbered. Explanation: After inlining, an unused
+   register is marked as "changed by the assembly" but can immediately be
+   re-used by the compiler, which is the same as a clobbered register. System
+   calls should generally be inlined -- and even if they aren't, the unused
+   return values will probably be passed in caller-saved registers (this is true
+   for the C ABI, so probably true for the Rust ABI), which are treated as
+   clobbered regardless.
+1. Command always needs to set all four registers, because otherwise we risk
+   leaking data to capsules (which are untrusted for confidentiality). Exit,
+   Memop, and Yield do not have this concern because they are implemented by the
+   core kernel (which is trusted).
+
+## Yield Method Breakdown
+
+`yield-wait` does not take any arguments, while `yield-no-wait` takes 1.
+Therefore, to avoid setting arguments unnecessarily, `yield-wait` and
+`yield-no-wait` should be distinct functions.
+
+## Remaining Method Breakdown
+
+After taking care of `yield`, we break down the remaining system calls by the
+number of arguments they take. Here is the breakdown (Exit and Memop are listed
+twice because they have sub-operations with varying number of arguments).
+
+* **Zero:**
+* **One:** Exit, Memop
+* **Two:** Exit, Memop
+* **Three:**
+* **Four:** Command, Read-Only Allow, Read-Write Allow, Subscribe
+
+Therefore we need three more methods: `one_arg_syscall`, `two_arg_syscall`,
+`four_arg_syscall`.
+
+## `u32` versus `usize`
+
+The decision of where to use `u32` and where to use `usize` can be a bit tricky.
+The Tock syscall ABI is currently only specified for 32-bit systems, so on real
+Tock systems both types match the size of a register, but the unit test
+environment can be either 32 bit or 64 bit. `RawSyscalls` uses `usize` for
+values that can contain pointers, so that pointers are not truncated in the unit
+test environment. To keep types as consistent as possible, it uses `u32` for
+register-sized values that cannot be pointers.
```

Comment:
I'm not sure I follow. Is there a reason you can't make these `usize` too, and let them get sign/zero extended? It feels a bit gross to have some be u32 and some be usize for what seems (from a trait reader's perspective) to be arbitrary reasons.

Reply:
I could, it would just result in an unnecessary pair of casts, as these are `u32`'s on both sides of the `RawSyscalls` abstraction.